sdks/python/apache_beam/io/filebasedsource.py

```diff
@@ -308,3 +313,102 @@ def read(self, range_tracker):
 
   def default_output_coder(self):
     return self._file_based_source.default_output_coder()
+
+
+class _ExpandIntoRanges(DoFn):
+
+  def __init__(
+      self, splittable, compression_type, desired_bundle_size, min_bundle_size):
+    self._desired_bundle_size = desired_bundle_size
+    self._min_bundle_size = min_bundle_size
+    self._splittable = splittable
+    self._compression_type = compression_type
+
+  def process(self, element, *args, **kwargs):
+    match_results = FileSystems.match([element])
+    for metadata in match_results[0].metadata_list:
+      splittable = (
+          self._splittable and
+          _determine_spelittability_from_compression_type(
+              metadata.path, self._compression_type))
+
+      if not splittable:
```

Comment:
Nit: Invert (put the not clause as the else clause). I would actually not even store this as an intermediate, i.e. just write "if self._spittable and ..."